src/App.tsx

```diff
@@ -203,6 +268,65 @@ class Board extends React.Component<BoardProps, BoardState> {
         this.setState(newState);
     }
 
+    public updateHistory(currentState: BoardState): BoardState {
+        const newState = _.cloneDeep(currentState);
+        const instanceToAdd: Instance = {
+            markingMap: _.cloneDeep(currentState.markingMap),
+            values: _.cloneDeep(currentState.values),
+        };
+        const oldInstance = newState.history[newState.historyIndex];
+        if (_.isEqual(instanceToAdd, oldInstance)) {
+            return newState;
+        }
+        newState.historyIndex = currentState.historyIndex + 1;
+        // On any history update the future history must be
+        // deleted because it is no longer the future of the
+        // new current (watch Back to the Future)
+        newState.history.splice(newState.historyIndex, currentState.history.length);
+        newState.history.push(instanceToAdd);
```

Comment:
0. Cone everything such that "modifications" are on new objects
1. Create your new instance such that its `prevInstance` is the old instance
2. "Modify" the old instance so that its `nextInstance` is the new instance
3. Revel in the fact that you're actually using intro to CS data structures in real life